File: julia/nbody.jl
using Random

const SOFTENING = 1e-9
const POSMAX = 100

mutable struct Body
    x::Int
    y::Int
    z::Int
    vx::Float64
    vy::Float64
    vz::Float64
    m::Float64
    key::Int
    
    function Body(x::Int, y::Int, z::Int, vx::Float64, vy::Float64, vz::Float64, m::Float64)
        new(x, y, z, vx, vy, vz, m, 0)
    end
    
    function Body(x::Int, y::Int, z::Int, vx::Float64, vy::Float64, vz::Float64, m::Float64, key::Int)
        new(x, y, z, vx, vy, vz, m, key)
    end
end

mutable struct Node
    key::Int
    isLeaf::Bool
    parent::Union{Node, Nothing}
    children::Vector{Union{Node, Nothing}}
    whichChildren::UInt8
    
    function Node(key::Int, isLeaf::Bool)
        new(key, isLeaf, nothing, fill(nothing, 8), 0)
    end
end

mutable struct Octree
    root::Node
    leafLength::Int
    nLevels::Int
    
    function Octree(maxKeyLength::Int)
        root = Node(1, false)
        leafLength = maxKeyLength
        nLevels = div(maxKeyLength, 3)
        println("Octree with $(nLevels) levels")
        new(root, leafLength, nLevels)
    end
end

mutable struct LCGRandomEngine
    state::UInt32
    
    function LCGRandomEngine(seed::Int)
        new(UInt32(seed))
    end
end

function Base.rand(rng::LCGRandomEngine)
    a = UInt32(1664525)
    c = UInt32(1013904223)
    
    rng.state = a * rng.state + c
    
    return rng.state / (1.0 * typemax(UInt32))
end

function Base.rand(rng::LCGRandomEngine, range::UnitRange{Int})
    width = range.stop - range.start + 1
    
    r = rand(rng) * width
    
    return range.start + floor(Int, r)
end

function getKeyNoPrepend(body::Body)
    key = 0
    
    x, y, z = body.x, body.y, body.z
    
    i = 0
    while x != 0 || y != 0 || z != 0
        if (body.x & (1 << i)) != 0
            key |= (1 << (3*i))
        end
        if (body.y & (1 << i)) != 0
            key |= (1 << (3*i + 1))
        end
        if (body.z & (1 << i)) != 0
            key |= (1 << (3*i + 2))
        end
        
        x >>= 1
        y >>= 1
        z >>= 1
        i += 1
    end
    
    return key
end

function binaryLength(n::Int)
    if n == 0
        return 0
    end
    
    length = floor(Int, log2(n)) + 1
    if length % 3 != 0
        length += 3 - length % 3
    end
    
    @assert length % 3 == 0
    return length
end

function binaryString(k::Int)
    if k <= 1
        return string(k % 2)
    else
        return binaryString(div(k, 2)) * string(k % 2)
    end
end

function addChild!(octree::Octree, parent::Node, index::Int, isLeaf::Bool, mykey::Int)
    @assert parent.isLeaf == false
    @assert mykey > 0
    
    parent.children[index+1] = Node(mykey, isLeaf)
    parent.children[index+1].parent = parent
    parent.whichChildren |= 1 << index
end

function insert!(octree::Octree, body::Body)
    current = octree.root
    key = body.key
    
    @assert key >> (octree.leafLength - 1) == 1
    @assert octree.nLevels * 3 == octree.leafLength - 1
    
    level = 0
    while level < octree.nLevels - 1
        shifted = key >> (3 * (octree.nLevels - level - 1))
        index = shifted & 0x7
        
        if current.children[index+1] === nothing
            addChild!(octree, current, index, false, shifted)
        end
        
        current = current.children[index+1]
        level += 1
    end
    
    @assert level == octree.nLevels - 1
    index = key & 0x7
    
    @assert current.children[index+1] === nothing
    addChild!(octree, current, index, true, key)
end

function printTree(node::Union{Node, Nothing}, level::Int)
    if node === nothing
        return
    end
    
    print("  " ^ level)
    println("0b$(binaryString(node.key)) ($(node.key))")
    
    if !node.isLeaf
        for i in 0:7
            if node.children[i+1] !== nothing
                printTree(node.children[i+1], level + 1)
            end
        end
    end
end

function bodyForceRange!(bodies::Vector{Body}, dt::Float64, start::Int, finish::Int, n::Int)
    for i in start:finish
        Fx, Fy, Fz = 0.0, 0.0, 0.0
        
        for j in 1:n
            dx = Float64(bodies[j].x - bodies[i].x)
            dy = Float64(bodies[j].y - bodies[i].y)
            dz = Float64(bodies[j].z - bodies[i].z)
            distSqr = dx^2 + dy^2 + dz^2 + SOFTENING
            invDist = 1.0 / sqrt(distSqr)
            invDist3 = invDist^3
            
            Fx += dx * bodies[j].m * invDist3
            Fy += dy * bodies[j].m * invDist3
            Fz += dz * bodies[j].m * invDist3
        end
        
        bodies[i] = Body(
            bodies[i].x, bodies[i].y, bodies[i].z,
            bodies[i].vx + dt * Fx, bodies[i].vy + dt * Fy, bodies[i].vz + dt * Fz,
            bodies[i].m, bodies[i].key
        )
    end
end

function bodyForce!(bodies::Vector{Body}, dt::Float64, n::Int)
    mid = div(n, 2)
    bodyForceRange!(bodies, dt, 1, mid, n)
    bodyForceRange!(bodies, dt, mid+1, n, n)
end

function integratePositionsRange!(bodies::Vector{Body}, dt::Float64, start::Int, finish::Int)
    for i in start:finish
        fx = bodies[i].x + bodies[i].vx * dt
        fy = bodies[i].y + bodies[i].vy * dt
        fz = bodies[i].z + bodies[i].vz * dt
        
        newX = Int(round(fx))
        newY = Int(round(fy))
        newZ = Int(round(fz))
        
        newX = clamp(newX, 0, POSMAX-1)
        newY = clamp(newY, 0, POSMAX-1)
        newZ = clamp(newZ, 0, POSMAX-1)
        
        bodies[i] = Body(
            newX, newY, newZ,
            bodies[i].vx, bodies[i].vy, bodies[i].vz,
            bodies[i].m, bodies[i].key
        )
    end
end

function integratePositions!(bodies::Vector{Body}, dt::Float64, n::Int)
    mid = div(n, 2)
    integratePositionsRange!(bodies, dt, 1, mid)
    integratePositionsRange!(bodies, dt, mid+1, n)
end

function nbody_simulation(nBodies::Int = 2)
    println("NBody using Octree")
    
    rng = LCGRandomEngine(2025)
    
    dt = 0.01
    nIters = 10
    
    println("Simulating $nBodies bodies")
    
    bodies = Vector{Body}(undef, nBodies)
    
    for i in 1:nBodies
        bodies[i] = Body(0, 0, 0, 0.0, 0.0, 0.0, 0.0, 0)
    end
    
    maxKeyLength = randomizeBodies(rng, bodies, nBodies)
    println("Randomized bodies with maxKeyLength = $maxKeyLength")
    
    octree = Octree(maxKeyLength)
    
    for i in 1:nBodies
        insert!(octree, bodies[i])
    end
    
    printTree(octree.root, 0)

    GC.gc()
    
    totalTime = 0.0
    
    for iter in 1:nIters
        startTime = time()
        
        bodyForce!(bodies, dt, nBodies)
        
        integratePositions!(bodies, dt, nBodies)
        
        newMaxKeyLength = 0
        for i in 1:nBodies
            body = bodies[i]
            key = getKeyNoPrepend(body)
            keyLength = binaryLength(key)
            
            if keyLength > newMaxKeyLength
                newMaxKeyLength = keyLength
            end
            
            bodies[i] = Body(
                body.x, body.y, body.z,
                body.vx, body.vy, body.vz,
                body.m, key
            )
        end
        
        prepend = 1 << newMaxKeyLength
        octree = Octree(newMaxKeyLength + 1)
        
        for i in 1:nBodies
            body = bodies[i]
            bodies[i] = Body(
                body.x, body.y, body.z,
                body.vx, body.vy, body.vz,
                body.m, body.key + prepend
            )
            
            insert!(octree, bodies[i])
        end
        
        endTime = time()
        tElapsed = endTime - startTime
        
        if iter > 1
            totalTime += tElapsed
        end
        
        println("Iteration $iter: $tElapsed seconds")
    end
    
    if nIters > 1
        avgTime = totalTime / (nIters - 1)
        println("Average time for iterations 2 through $nIters: $avgTime seconds.")
        println("$nBodies Bodies: average $(1e-9 * nBodies * nBodies / avgTime) Billion Interactions / second")
    end
end

function randomizeBodies(rng::LCGRandomEngine, bodies::Vector{Body}, n::Int)
    maxKeyLength = 0
    
    for i in 1:n
        x = rand(rng, 0:POSMAX-1)
        y = rand(rng, 0:POSMAX-1)
        z = rand(rng, 0:POSMAX-1)
        
        tempBody = Body(x, y, z, 0.0, 0.0, 0.0, 0.0)
        
        key = getKeyNoPrepend(tempBody)
        keyLength = binaryLength(key)
        
        if keyLength > maxKeyLength
            maxKeyLength = keyLength
        end
        
        vx = rand(rng) * 2 - 1
        vy = rand(rng) * 2 - 1
        vz = rand(rng) * 2 - 1
        m = rand(rng) + 0.1
        
        bodies[i] = Body(x, y, z, vx, vy, vz, m, key)
    end
    
    prepend = 1 << maxKeyLength
    println("Using prepend value: $prepend (2^$maxKeyLength)")
    
    for i in 1:n
        bodies[i] = Body(
            bodies[i].x, bodies[i].y, bodies[i].z,
            bodies[i].vx, bodies[i].vy, bodies[i].vz,
            bodies[i].m, bodies[i].key + prepend
        )
    end
    
    return maxKeyLength + 1
end

function main()
    nBodies = 2
    if length(ARGS) > 0
        nBodies = parse(Int, ARGS[1])
    end
    try
        nbody_simulation(nBodies)
    catch e
        println("Error in simulation: $e")
        println(catch_backtrace())
    end
end

main()

File: cpp/nbody.cpp
#include "ctimer.h"
#include "octree.h"
#include "philox_engine.h"
#include <assert.h>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <random>
// #include <cilk.h>

#define SOFTENING 1e-9f
#define POSMAX 100

bool USE_TREE = false; // use tree or list
bool USE_BH = true;    // use barnes hut or not
bool PRINT_TIME = true;
float MAC_PARAM = .5; // MAC parameter

bool MAC(float target_x, float target_y, float target_z, float x, float y,
         float z, float mass) {

  float distance =
      sqrtf((target_x - x) * (target_x - x) + (target_y - y) * (target_y - y) +
            (target_z - z) * (target_z - z));
  float source_mass = mass;

  bool result = distance > MAC_PARAM;

  return distance / source_mass > MAC_PARAM;
}

// Randomize the positions and velocities in the range [-1, 1],
// and assign a mass (every 7th float) in the range [0.1, 1.0].
int randomizeBodies(PhiloxEngine &rng, Body *bodies, int n) {
  int maxKeyLength = 0;
  std::uniform_real_distribution<float> x_dis(0, 1);
  std::uniform_real_distribution<float> v_dis(0, 1);

  for (int i = 0; i < n; i++) {
    Body &body = bodies[i];
    body.x = x_dis(rng);
    body.y = x_dis(rng);
    body.z = x_dis(rng);

    body.index = i;

    body.key = getKeyNoPrepend(body);
    int keylength = binaryLength(body.key);
    if (keylength > maxKeyLength) {
      maxKeyLength = keylength;
    }

    body.vx = (v_dis(rng)) * 2 - 1;
    body.vy = (v_dis(rng)) * 2 - 1;
    body.vz = (v_dis(rng)) * 2 - 1;
    body.m = (v_dis(rng)) + 0.1f;
  }

  // prepend all keys
  int prepend = 1 << maxKeyLength;
  for (int i = 0; i < n; i++) {
    bodies[i].key += prepend;
  }

  return maxKeyLength + 1;
}

int MACInteractionsDFT(std::vector<DFTNode> dft, Body *bodies, int target,
                       float dt, int nBodies) {
  // iterate over all bodies (targets)
  int nInteractions = 0;
  float Fx = 0.0f, Fy = 0.0f, Fz = 0.0f;
  for (int j = 0; j < dft.size(); j++) {
    if (dft[j].isLeaf) {
      for (int bIndex : dft[j].bodies) {
        nInteractions++;
        float dx = bodies[bIndex].x - bodies[target].x;
        float dy = bodies[bIndex].y - bodies[target].y;
        float dz = bodies[bIndex].z - bodies[target].z;
        float distSqr = dx * dx + dy * dy + dz * dz + SOFTENING;
        float invDist = 1.0f / sqrtf(distSqr);
        float invDist3 = invDist * invDist * invDist;

        Fx += dx * bodies[bIndex].m * invDist3;
        Fy += dy * bodies[bIndex].m * invDist3;
        Fz += dz * bodies[bIndex].m * invDist3;
      }
    } else if (MAC(bodies[target].x, bodies[target].y, bodies[target].z,
                   dft[j].x, dft[j].y, dft[j].z, dft[j].mass)) {
      nInteractions++;
      float dx = dft[j].x - bodies[target].x;
      float dy = dft[j].y - bodies[target].y;
      float dz = dft[j].z - bodies[target].z;
      float distSqr = dx * dx + dy * dy + dz * dz + SOFTENING;
      float invDist = 1.0f / sqrtf(distSqr);
      float invDist3 = invDist * invDist * invDist;

      Fx += dx * dft[j].mass * invDist3;
      Fy += dy * dft[j].mass * invDist3;
      Fz += dz * dft[j].mass * invDist3;
      j = dft[j].autorope - 1; // -1 because we increment j in the for loop
    }
  }

  bodies[target].vx += dt * Fx;
  bodies[target].vy += dt * Fy;
  bodies[target].vz += dt * Fz;
  return nInteractions;
}

int allInteractionsDFT(std::vector<DFTNode> dft, Body *bodies, int target,
                       float dt, int nBodies) {
  // iterate over all bodies (targets)
  int nInteractions = 0;
  float Fx = 0.0f, Fy = 0.0f, Fz = 0.0f;

  for (int j = 0; j < dft.size(); j++) {
    if (dft[j].isLeaf) {
      for (int bIndex : dft[j].bodies) {
        nInteractions++;
        float dx = bodies[bIndex].x - bodies[target].x;
        float dy = bodies[bIndex].y - bodies[target].y;
        float dz = bodies[bIndex].z - bodies[target].z;
        float distSqr = dx * dx + dy * dy + dz * dz + SOFTENING;
        float invDist = 1.0f / sqrtf(distSqr);
        float invDist3 = invDist * invDist * invDist;

        Fx += dx * bodies[bIndex].m * invDist3;
        Fy += dy * bodies[bIndex].m * invDist3;
        Fz += dz * bodies[bIndex].m * invDist3;
      }
    }
  }
  bodies[target].vx += dt * Fx;
  bodies[target].vy += dt * Fy;
  bodies[target].vz += dt * Fz;
  return nInteractions;
}

void allInteractionsDS(Body *bodies, float dt, int nBodies) {
  int n = nBodies;

  for (int i = 0; i < n; ++i) {

    float Fx = 0.0;
    float Fy = 0.0;
    float Fz = 0.0;

    // compute force, update velocities
    for (int j = 0; j < n; ++j) {

      float dx = bodies[j].x - bodies[i].x;
      float dy = bodies[j].y - bodies[i].y;
      float dz = bodies[j].z - bodies[i].z;
      float distSqr = dx * dx + dy * dy + dz * dz + SOFTENING;
      float invDist = 1.0f / sqrtf(distSqr);
      float invDist3 = invDist * invDist * invDist;

      Fx += dx * bodies[j].m * invDist3;
      Fy += dy * bodies[j].m * invDist3;
      Fz += dz * bodies[j].m * invDist3;
    }

    // update velocity
    bodies[i].vx += dt * Fx;
    bodies[i].vy += dt * Fy;
    bodies[i].vz += dt * Fz;
  }
}

void bodyForce(Body *p, float dt, int n, std::vector<DFTNode> dft) {

  if (USE_TREE && USE_BH) {
  } else if (USE_TREE) {
  } else {
    allInteractionsDS(p, dt, n);
  }
  // // printf("DFT size: %ld\n", dft.size());
  // for (int i = 0; i < n; i++) {
  //   if (USE_TREE && USE_BH)
  //     MACInteractionsDFT(dft, p, i, dt, n);
  //   else if (USE_TREE)
  //     allInteractionsDFT(dft, p, i, dt, n);
  //   else
  //     allInteractionsDS(p, i, dt, n);
  // }
  //
  // printf("Total interactions: %ld\n", totalInteractions);
}

void integratePositions(Body *p, float dt, int start, int end) {
  for (int i = start; i < end; i++) {
    p[i].x += (p[i].vx * dt);
    p[i].y += (p[i].vy * dt);
    p[i].z += (p[i].vz * dt);
  }
}

void nbodyIterate(Body *p, float dt, int nBodies, int nIters,
                  int maxkeylength) {
  double totalTime = 0.0;
  ctimer_t timer;
  for (int iter = 1; iter <= nIters; iter++) {

    std::vector<DFTNode> dft;

    Octree *octree;

    ctimer_start(&timer);

    // Reconstruct octree in every timestep
    if (USE_TREE) {
      octree = new Octree(maxkeylength);
      int nUniqueLeaves = 0;
      for (int i = 0; i < nBodies; i++) {
        nUniqueLeaves += octree->insert(p[i]);
      }

      octree->buildDFT(dft, p);
      printf("Octree built with %d buckets (leaves with unique key)\n",
             nUniqueLeaves);
      // octree->printTree(octree->root, 0);
    }

    bodyForce(p, dt, nBodies, dft);

    if (USE_TREE)
      delete octree;

    // Integrate positions
    integratePositions(p, dt, 0, nBodies);

    ctimer_stop(&timer);
    ctimer_measure(&timer);

    double tElapsed = timespec_sec(timer.elapsed);

    // skip the first iteration as warm-up
    if (iter > 1)
      totalTime += tElapsed;
  }

  double avgTime = totalTime / (double)(nIters - 1);

  if (PRINT_TIME) {
    printf("Total time = %.3f seconds, Average time = %.3f seconds.\n",
           totalTime, avgTime);
    printf("%d Bodies: average %0.3f Billion Interactions / second\n", nBodies,
           1e-9 * nBodies * nBodies / avgTime);
  }
}

void printBodiesToFile(Body *p, int nBodies) {
  FILE *file;
  if (USE_TREE)
    if (USE_BH)
      file = fopen("bodiesAfterBH.txt", "a");
    else
      file = fopen("bodiesAfterDFT.txt", "a");
  else
    file = fopen("bodiesAfterDS.txt", "a");

  if (file) {
    for (int i = 0; i < nBodies; i++) {
      fprintf(file, "Body %d: x=%f, y=%f, z=%f\n", i, p[i].x, p[i].y, p[i].z);
    }
    fclose(file);
  } else {
    printf("Error opening file for writing.\n");
  }
}

void checkAccuracy(Body *p, Body *orig, int nBodies) {
  printf("-----------------------------------------------\n");
  printf("ACCURACY CHECK AGAINST DS\n");

  USE_TREE = false;
  USE_BH = false;
  PRINT_TIME = false;
  nbodyIterate(orig, 0.01f, nBodies, 10, 0);
  PRINT_TIME = true;
  // Compute RMS error(standard check for positional accuracy)
  float rmsError = 0.0f;
  for (int i = 0; i < nBodies; i++) {

    float dx = p[i].x - orig[i].x;
    float dy = p[i].y - orig[i].y;
    float dz = p[i].z - orig[i].z;
    rmsError += dx * dx + dy * dy + dz * dz;
  }
  rmsError = std::sqrt((rmsError / nBodies));
  printf("RMS error: %f\n", rmsError);
}

using namespace std;

int main(const int argc, const char **argv) {
  printf("NBody Simulation\n");

  int nBodies = 2000;
  if (argc > 1)
    nBodies = atoi(argv[1]);

  if (argc > 2) {
    nBodies = atoi(argv[1]);
    const char *method = argv[2];
    if (strcmp(method, "DS") == 0) {
      USE_TREE = false;
      USE_BH = false;
    } else if (strcmp(method, "DFT") == 0) {
      USE_TREE = true;
      USE_BH = false;
    } else if (strcmp(method, "MAC") == 0) {
      USE_TREE = true;
      USE_BH = true;
    } else {
      printf("Unknown method: %s\n", method);
      return -1;
    }
  }

  if (argc > 3) {
    MAC_PARAM = atof(argv[3]);
  }

  // TODO: other parameters we might want to control:
  // - SHIFT_DIGITS (controls how many keys/buckets are created)
  // - MAC_PARAM (controls the MAC parameter)
  // - SOFTENING (controls the softening factor)

  printf(" - %s method \n - %d bodies\n",
         USE_TREE ? (USE_BH ? "MAC" : "DFT") : "DS", nBodies);
  printf("-----------------------------------------------\n");

  const float dt = 0.01f; // time step
  const int nIters = 10;  // simulation iterations

  // Allocate memory for nBodies
  Body *particles = new Body[nBodies];

  // Initialize positions, velocities, and masses (7 values per body)
  PhiloxEngine rng(2025);
  int maxkeylength = randomizeBodies(rng, particles, nBodies);

  // make a copy of bodies for accuracy check
  Body *particles_cp = new Body[nBodies];
  for (int i = 0; i < nBodies; i++) {
    particles_cp[i] = particles[i];
  }

  nbodyIterate(particles, dt, nBodies, nIters, maxkeylength);
  checkAccuracy(particles, particles_cp, nBodies);

  delete[] particles;
  delete[] particles_cp;
  return 0;
}
